flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java

```diff
@@ -51,20 +53,20 @@
 
 	private static final Logger LOG = LoggerFactory.getLogger(DefaultExecutionTopology.class);
 
-	private final ExecutionGraph executionGraph;
-
 	private final Map<ExecutionVertexID, DefaultExecutionVertex> executionVerticesById;
 
 	private final List<DefaultExecutionVertex> executionVerticesList;
 
 	private final Map<IntermediateResultPartitionID, DefaultResultPartition> resultPartitionsById;
 
-	private final Map<ExecutionVertexID, DefaultSchedulingPipelinedRegion> pipelinedRegionsByVertex;
+	@Nullable
+	private Map<ExecutionVertexID, DefaultSchedulingPipelinedRegion> pipelinedRegionsByVertex;
 
-	private final List<DefaultSchedulingPipelinedRegion> pipelinedRegions;
+	@Nullable
+	private List<DefaultSchedulingPipelinedRegion> pipelinedRegions;
 
-	public DefaultExecutionTopology(ExecutionGraph graph) {
-		this.executionGraph = checkNotNull(graph, "execution graph can not be null");
+	private DefaultExecutionTopology(ExecutionGraph graph) {
```

Comment:
If we are already factoring out `initializePipelinedRegions`, could we generate all dependencies of `DefaultExecutionTopology` in the factory method `fromExecutionGraph`? This should generally separate concerns and make potential testing simpler.

I also would put this commit before the actual change of this PR if this is not much effort.

Reply:
I have reordered the commits to let the hotfix happen first.
But I do not feel like to factor out vertices and results creation out from `DefaultExecutionTopology` because I think it is the core part of `DefaultExecutionTopology`. One can use `TestingSchedulingTopology` to generate a controllable topology for testing.

Reply:
I meant testing `DefaultExecutionTopology` itself and its construction process separately
but it seems that tests have already been written for the constructor.
In this case, I would keep it as it was with final fields at least (it provides immutability and better guarantees)
if you think that construction does not need to be separated from the `DefaultExecutionTopology` as just a data holder.

Reply:
We cannot make `pipelinedRegions` final because it is built from a `Topology`.
And we should also avoid the dangerous behavior which published `this` in the constructor.

Reply:
As I understand, the topology is computed from vertices.
I meant this refactoring: https://github.com/azagrebin/flink/commit/9bd51df78208725a825239bfe16aa3e9ae296c4f.

Reply:
The topology consists of vertices, results and edges. `DefaultExecutionTopology` currently built these things in the constructor. I feel that moving them into the factory method does not bring any immediate benefit but instead introduces risk. So I prefer to do it only when it is needed.

Reply:
Ok, I think we can merge the PR as is and then I can open a follow-up PR right-away with my.
My concern is the private setter as a compromise to work on a partially constructed Topology.
I still believe that having a separate step-by-step dependencies generation with immutable data structures is more clear in future. There should be also no risks if tests pass.